src/java.base/share/classes/sun/util/locale/BaseLocale.java

```diff
@@ -90,6 +92,9 @@ public final class BaseLocale {
         }
     }
 
+    // Non-normalized to normalized BaseLocale cache for saving costly normalizations
+    private static final Map<BaseLocale, BaseLocale> CACHE = ReferencedKeyMap.create(true, ConcurrentHashMap::new);
```

Comment:
You can now simply use a `ReferencedKeySet<BaseLocale>` and call [`intern(T, UnaryOperator<T>)`](https://github.com/openjdk/jdk/blob/ce2a7ea40a22c652e5f8559c91d5eea197e2d708/src/java.base/share/classes/jdk/internal/util/ReferencedKeySet.java#L202) to cache a normalized variety computed from a non-normalized base.

Reply:
I need to double-check that it won't cause performance regression, will be visiting this later.